Add in support for casting decimal to other number types

revans2 opened the pull request:
This fixes #3610. I mistakenly told @ttnghia that we supported this already, but I was wrong so I am adding it in.  Not sure if this good for 21.10 or not because it was filed as a bug, but it is kind of new work also.

revans2 commented:
build

jlowe reviewed (commented):
I see changes only to the 320 and 311until320-nondb shims.  Seems like we're missing some of the shims?

revans2 commented:
> I see changes only to the 320 and 311until320-nondb shims. Seems like we're missing some of the shims?

looks like I need to upmerge. I grepped before, but there are new ones that I missed

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala:
```diff
@@ -355,20 +358,44 @@ object GpuCast extends Arm {
+        assertValuesInRange(input, Scalar.fromDecimal(min), Scalar.fromDecimal(max))
+        if (dt.precision <= DType.DECIMAL32_MAX_PRECISION && dt.scale < 0) {
+          // This is a work around for https://github.com/rapidsai/cudf/issues/9281
+          withResource(input.castTo(DType.create(DType.DTypeEnum.DECIMAL64, -dt.scale))) { tmp =>
```
Mostly curiosity, the issue linked doesn't mention negative scales, so why do we need to negate the scale here again?

revans2 replied:
cudf scale = - spark scale

abellina replied:
thanks!

revans2 commented:
build

revans2 commented:
@jlowe I have addressed your review comments, but I am seeing some build errors on cdh and 313.  Not 100% sure what is happening but @tgravescs is looking into it because it appears to be on the base branch now.

jlowe commented on sql-plugin/src/main/311until320-nondb/scala/com/nvidia/spark/rapids/shims/v2/SparkBaseShims.scala:
```diff
@@ -165,7 +165,7 @@ abstract class SparkBaseShims extends Spark31XShims {
         // stringChecks are the same
         // binaryChecks are the same
-        override val decimalChecks: TypeSig = DECIMAL_64 + STRING
+        override val decimalChecks: TypeSig = gpuNumeric + STRING
```
It looks like the shims are overriding this to the same behavior as the base method.  Would it be a better change to stop overriding the base version in the shims that don't need to change the behavior, or was this intentionally overridden redundantly?

revans2 replied:
Some of this was done because Spark itself changed what could be cast (dropped some things that just returned nulls). So at a minimum we had to override `sparkDecimalSig` but I thought it was cleaner to keep it all together and included `decimalChecks` with it.
